lib/launcher/src/Cardano/Launcher.hs

```diff
@@ -148,59 +156,70 @@ instance Exception ProcessHasExited
 withBackendProcessHandle
-    :: Tracer IO LauncherLog
+    :: MonadUnliftIO m
+    => Tracer m LauncherLog
     -- ^ Logging
     -> Command
     -- ^ 'Command' description
-    -> (Maybe Handle -> ProcessHandle -> IO a)
+    -> (Maybe Handle -> ProcessHandle -> m a)
     -- ^ Action to execute while process is running.
-    -> IO (Either ProcessHasExited a)
+    -> m (Either ProcessHasExited a)
 withBackendProcessHandle tr (Command name args before std_in std_out) action =
-    before >> withBackendCreateProcess tr process action
+    liftIO before >> withBackendCreateProcess tr process action
   where
-    process = (proc name args)  { std_in, std_out, std_err = std_out }
-
+    process = (proc name args) { std_in, std_out, std_err = std_out }
 
 -- | A variant of 'withBackendProcess' which accepts a general 'CreateProcess'
 -- object.
+--
+-- This function should ensure:
+-- 1. If the action finishes or throws an exception,
+--    then the process is also terminated.
+-- 2. If the process exits, then the action is cancelled.
 withBackendCreateProcess
-    :: Tracer IO LauncherLog
+    :: forall m a. (MonadUnliftIO m)
+    => Tracer m LauncherLog
     -- ^ Logging
     -> CreateProcess
     -- ^ 'Command' description
-    -> (Maybe Handle -> ProcessHandle -> IO a)
+    -> (Maybe Handle -> ProcessHandle -> m a)
     -- ^ Action to execute while process is running.
-    -> IO (Either ProcessHasExited a)
+    -> m (Either ProcessHasExited a)
 withBackendCreateProcess tr process action = do
     traceWith tr $ MsgLauncherStart name args
-    res <- fmap join $ tryJust spawnPredicate $
-        withCreateProcess process $ \mstdin _ _ h -> do
-            pid <- maybe "-" (T.pack . show) <$> getPid h
-            let tr' = contramap (WithProcessInfo name pid) tr
-            traceWith tr' MsgLauncherStarted
-
-            let waitForExit =
-                    ProcessHasExited name <$> interruptibleWaitForProcess tr' h
-            let runAction = do
-                    traceWith tr' MsgLauncherAction
-                    action mstdin h `finally` traceWith tr' MsgLauncherCleanup
-
-            race waitForExit runAction
-    either (traceWith tr . MsgLauncherFinish) (const $ pure ()) res
+    res <- withCreateProcessMod $ \mstdin ph -> do
+        pid <- maybe "-" (T.pack . show) <$> liftIO (getPid ph)
+        let tr' = contramap (WithProcessInfo name pid) tr
+        traceWith tr' MsgLauncherStarted
+        let waitForExit = ProcessHasExited name <$>
+                interruptibleWaitForProcess tr' ph
+        let runAction = bracket_
+                (traceWith tr' MsgLauncherAction)
+                (traceWith tr' MsgLauncherCleanup)
+                (action mstdin ph)
+        race waitForExit runAction
+
+    whenLeft res (traceWith tr . MsgLauncherFinish)
     pure res
   where
+    withCreateProcessMod inner =
+        fmap join $ tryJust spawnPredicate $
+        bracket (createProcess process) (liftIO . cleanupProcess) $
```

Comment:
Why not stick with: https://hackage.haskell.org/package/unliftio-0.2.13/docs/UnliftIO-Process.html#v:withCreateProcess :thinking: ?

Reply:
I need to revert this - it's a partial change.